csvw/csvw.go

```diff
@@ -0,0 +1,246 @@
+type Publisher struct {
+	Name string `json:"name,omitempty"`
+	Type string `json:"@type,omitempty"`
+	ID   string `json:"@id"` //a URL where more info is available
+}
+
+//Columns provides the nested structure expected within the tableSchema of a CSVW
+type Columns struct {
+	C     []Column `json:"columns"`
+	About string   `json:"aboutUrl"`
+}
+
+//Column provides the ability to define the JSON tags required specific
+//to each column within the CSVW
+type Column map[string]interface{}
+
+//Note can include alerts, corrections or usage notes which exist in the
+//dataset metadata and help describe the contents of the CSV
+type Note struct {
+	Type       string `json:"type"` // is this an enum?
+	Target     string `json:"target,omitempty"`
+	Body       string `json:"body"`
+	Motivation string `json:"motivation,omitempty"` // how is this different from type? do we need this? is this an enum?
+}
+
+var invalidHeader = errors.New("invalid header row - no V4_X cell")
+var missingDimensions = errors.New("no dimensions in provided metadata")
+
+//New CSVW returned with top level fields populated based on provided metadata
+func New(m *dataset.Metadata, csvURL string) *CSVW {
+	csvw := &CSVW{
+		Context:     "http://www.w3.org/ns/csvw",
+		Title:       m.Title,
+		Description: m.Description,
+		Issued:      m.ReleaseDate,
+		Theme:       m.Theme,
+		License:     m.License,
+		Frequency:   m.ReleaseFrequency,
+		URL:         csvURL,
+	}
+
+	for _, c := range m.Contacts {
+		csvw.Contact = append(csvw.Contact, Contact{
+			Name:      c.Name,
+			Telephone: c.Telephone,
+			Email:     c.Email,
+		})
+	}
+
+	if m.Publisher != nil {
+		csvw.Publisher = Publisher{
+			Name: m.Publisher.Name,
+			Type: m.Publisher.Type,
+			ID:   m.Publisher.URL,
+		}
+	}
+
+	return csvw
+}
+
+//Generate the CSVW structured metadata file to describe a CSV
+func Generate(metadata *dataset.Metadata, header, downloadURL, aboutURL string) ([]byte, error) {
+	if len(metadata.Dimensions) == 0 {
+		return nil, missingDimensions
+	}
+
+	h, offset, err := splitHeader(header)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("header split for CSVW generation", log.Data{"header": header, "column_offset": strconv.Itoa(offset)})
+
+	csvw := New(metadata, downloadURL)
+
+	var list []Column
+	obs := newObservationColumn(downloadURL, h[0], metadata.UnitOfMeasure)
+	list = append(list, obs)
+	log.Info("added observation column to CSVW", log.Data{"column": obs})
+
+	//add data markings columns
+	if offset != 0 {
+		for i := 1; i <= offset; i++ {
+			c := newColumn(i, downloadURL, h[i], "")
+			list = append(list, c)
+			log.Info("added data marking column to CSVW", log.Data{"column": c})
+		}
+	}
+
+	offset++
+	h = h[offset:]
+
+	//add dimensions columns
+	for i := 0; i < len(h); i = i + 2 {
+		c, l := newCodeAndLabelColumns(offset, i, downloadURL, h, metadata.Dimensions)
+		log.Info("added dimension columns to CSVW", log.Data{"code_column": c, "label_column": l})
```

Comment:
This log line would have added the one dimension column, so should it not be:

`log.Info("added dimension column to CSVW", log.Data{"code_column": c, "label_column": l})`

Reply:
No because its adding two, I could change it to 'added pair of dimension columns to CSVW' if that's clearer?